apps/lpasswd.1
.\" A man page for lpasswd
.\" Copyright (C) 2005 Red Hat, Inc.
.\"
.\" This is free software; you can redistribute it and/or modify it under
.\" the terms of the GNU Library General Public License as published by
.\" the Free Software Foundation; either version 2 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU Library General Public
.\" License along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\" Author: Miloslav Trmac <mitr@redhat.com>
.TH lpasswd 1 "Jan 12 2005" libuser

.SH NAME
lpasswd \- Change group or user password

.SH SYNOPSIS
lpasswd [\fIOPTION\fR]... [\fIname\fR]

.SH DESCRIPTION
Changes password of user or group \fIname\fR.

If the \fIname\fR argument is not provided,
username of the invoking user is used;
the \fIname\fR argument is ignored if
.B lpasswd
is run set-uid to a different user.

Entering an empty password (by pressing Enter)
aborts the password setting operation.

.SH OPTIONS
.TP
\fB\-F\fR, \fB\-\-plainpassword-fd\fR=\fIfd\fR
Read password from file descriptor \fIfd\fR, terminated by end of file,
\fB'\\r'\fR or \fB'\\n'\fR.

.TP
\fB\-f\fR, \fB\-\-password-fd\fR=\fIfd\fR
Read password hash from file descriptor \fIfd\fR, terminated by end of file,
\fB'\\r'\fR or \fB'\\n'\fR.

.TP
\fB\-g\fR, \fB\-\-group\fR
Change password of group \fIname\fR.
By default the password of user \fIname\fR is changed.

.TP
\fB\-i\fR, \fB\-\-interactive\fR 
Ask all questions when connecting to the user database,
even if default answers are set up in
.B libuser
configuration.

.TP
\fB\-P\fR, \fB\-\-plainpassword\fR=\fIpassword\fR
Set the password to \fIpassword\fR.
Note that the password can be viewed while running
.BR lpasswd
using tools such as
.BR ps (1)\fR.

.TP
\fB\-p\fR, \fB\-\-password\fR=\fIencrypted\fR
Set the password to the password represented by the hash \fIencrypted\fR.
Note that the hash can be viewed while running
.BR lpasswd
using tools such as
.BR ps (1)\fR.

.SH EXIT STATUS
The exit status is 0 on success, nonzero on error.


NEWS
0.56.9
* Warn in lusermod if changing a primary group ID to a group that does not
  exist. (#1)
* Fix pastos in man pages.

0.56.8
* New home page at https://fedorahosted.org/libuser/ .

0.56.7
* Fix a crash with disabled SELinux
* Add support for SHA256 and SHA512 in password hashes.
* Fix file locking on some architectures
* Remove default.-c, moving the provided functions to libuser proper.

0.56.6
* Set SELinux file contexts when creating home directories and preserve them
  when moving home directories

0.56.5
* Work around spurious error messages when run against the Fedora Directory
  server
* Fix error reporting when creating home directories and creating / removing
  mail spool files

0.56.4
* Update the last password change date field when changing passwords.

0.56.3
* Allow specifying a SASL mechanism (#240904, original patch by Simo Sorce).

0.56.2
* Update translations

0.56.1
* When changing passwords, only silently ignore known shadow markers, not all
  invalid hashes

0.56
* Document the correct types used for attribute values.  Use these types for
  parsing, to avoid corrupting number-like strings, e.g. '07' -> 7; this
  expands the API requirements
* Miscellaneous bug fixes, optimizations and cleanups; module interface ABI
  has changed

0.55
* Remove the quota library and Python module.  It doesn't even compile and
  has no known users.
* Add support for the 64-bit API of Python 2.5
* Minor cleanups

0.54.8
* Add importing of HOME from default/useradd.

0.54.7
* Update translations

0.54.6
* Fix bugs in handling of invalid lines in the files and shadow modules.
* Fix pattern matching in lu_*_enumerate_full in the files and shadow modules.
* Add more error reporting, return non-zero exit status on error from utils.
* Use the skeleton directory specified in libuser.conf by Python
  admin.createHome and admin.addUser, add parameter skeleton= to admin.addUser.

0.54.5
* Don't reference @pkglibdir@ in libuser.conf.5 to avoid multilib file
  conflicts.

0.54.4
* Fix compilation with pre-C99 compilers (#179385, patch by Dan Yefimov).
* Allow building without Python (#179384, original patch by Dan Yefimov).

0.54.3
* Fix a crash when lpasswd is run without specifying an user name

0.54.2
* Avoid using deprecated openldap functions

0.54.1
* Support for importing of configuration from shadow (/etc/login.defs and
  /etc/default/useradd)
* New libuser.conf(5) man page
* Minor cleanups and bug fixes all over the code

0.54
* Make sure attributes with no values can never appear
* Fix crash in the "files" module when an attribute is missing
* Use hidden visibility for internal functions, remove them from
  libuser/user_private.h; this changes module interface ABI
* Miscellaneous source code simplifications

0.53.8
* Permit "portable" user and group names as defined by SUSv3, plus trailing $
* Disable building static libraries by default
* Miscellaneous build machinery improvements

0.53.7
* Add missing translations
* Update translations

0.53.6
* Allow empty configuration values.

0.53.5
* Ignore nss_compat lines in the "files" module.
* Autodetect Python version.

0.53.4
* Fix adding of objectclasses to LDAP user accounts.

0.53.3
* Handle more I/O failures.

0.53.2
* Important bug fixes in lchage, lgroupmod, lnewusers and lusermod; minor bug
  fixes in lpasswd and luseradd.
* Add man pages for the utilities.

0.53.1
* Export UT_NAMESIZE from <utmp.h> to Python

0.53
* Support UID and GID values larger than LONG_MAX (#124967)
* Fix updating of groups after user renaming in lusermod
* Allow setting a shadow password even if the current shadow password is
  invalid (#131180)
* Add lu_{user,group}_unlock_nonempty (#86414); module interface ABI has
  changed
* Miscellaneous bug and memory leak fixes

0.52.6
* Mark more strings for translation
* Make error reporting more consistent and more verbose, output error messages
  on stderr.
* Port sasldb backend to Cyrus SASL v2, make it at least minimally usable

0.52.5
* Fix home directory renaming in ADMIN.modifyUser (#135280)
* Further Python reference counting fixes

0.52.4
* Memory leak fixes

0.52.3
* Fix compilation without libuser headers already installed (#134085)

0.52.2
* Allow LDAP connection using ldaps, custom ports or without TLS (original
  patch from Pawel Salek).

0.52.1
* Fix freecon() of uninitialized value in files/shadow module

0.52
* Usable LDAP backend
* Miscellaneous bug fixes

0.51.12
* Don't claim success and exception at the same time (#133479)
* LDAP fixes, second round
* Various other bug fixes

0.51.11
* Allow documented optional arguments in Python
  ADMIN.{addUser,modifyUser,deleteUser} (#119812)
* Add man pages for lchfn and lchsh
* LDAP fixes, first round
* Avoid file conflict on multilib systems
* Call ldconfig correctly

0.51.10
* Don't attempt to lookup using original entity name after entity
  modification (rename in particular) (#78376, #121252)
* Fix copying of symlinks from /etc/skel (#87572, original patch from gLaNDix)
* Make --enable-quota work, and fix the quota code to at least compile (#89114)
* Fix several bugs (#120168, original patch from Steve Grubb)
* Don't hardcode python version in spec file (#130952, from Robert Scheck)
* Properly integrate the SELinux patch, it should actually be used now, even
  though it was "enabled" since 0.51.7-6

0.51.9
* Fix various typos
* Document library interfaces
* Build all shared libraries with -fPIC (#72536)

0.51.8
* Update to build with latest autotools and gtk-doc
* Update ALL_LINGUAS and POTFILES.in
* Rebuild to depend on newer openldap

0.51.7-7
* fix is_selinux_enabled call

0.51.7-3
* Add SELinux support

0.51.7
* ldap: set error codes correctly when we encounter failures initializing
* don't double-close modules which fail initialization
* ldap: don't set an error in cases where one is already set

0.51.6
* use a crypt salt consistent with the defaults/crypt_style setting when
  setting new passwords (#79337)

0.51.5

apps/lpasswd.c
/*
 * Copyright (C) 2000-2002 Red Hat, Inc.
 *
 * This is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Library General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#include <config.h>
#include <libintl.h>
#include <limits.h>
#include <locale.h>
#include <popt.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include "../lib/user.h"
#include "apputil.h"

int
main(int argc, const char **argv)
{
	struct lu_context *ctx;
	struct lu_ent *ent;
	struct lu_error *error = NULL;
	char *password = NULL, *cryptedPassword = NULL;
	const char *user;
	int c;
	int plain_fd = -1, crypted_fd = -1;
	int interactive = 0, groupflag = 0;
	poptContext popt;
	gboolean is_crypted;
	struct poptOption options[] = {
		{"interactive", 'i', POPT_ARG_NONE, &interactive, 0,
		 "prompt for all information", NULL},
		{"group", 'g', POPT_ARG_NONE, &groupflag, 0,
		 "set group password instead of user password", NULL},
		{"plainpassword", 'P', POPT_ARG_STRING, &password, 0,
		 "new plain password", NULL},
		{"password", 'p', POPT_ARG_STRING, &cryptedPassword, 0,
		 "new crypted password", NULL},
		{"plainpassword-fd", 'F', POPT_ARG_INT, &plain_fd, 0,
		 "read new plain password from given descriptor", NULL},
		{"password-fd", 'f', POPT_ARG_INT, &crypted_fd, 0,
		 "read new crypted password from given descriptor", NULL},
		POPT_AUTOHELP POPT_TABLEEND
	};

	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	setlocale(LC_ALL, "");

	popt = poptGetContext("lpasswd", argc, argv, options, 0);
	poptSetOtherOptionHelp(popt, _("[OPTION...] user"));
	c = poptGetNextOpt(popt);
	if (c != -1) {
		fprintf(stderr, _("Error parsing arguments: %s.\n"),
			poptStrerror(c));
		poptPrintUsage(popt, stderr, 0);
		exit(1);
	}
	user = poptGetArg(popt);

	if ((user == NULL) || (geteuid() != getuid())) {
		struct passwd *pwd;
		pwd = getpwuid(getuid());
		if (pwd != NULL) {
			fprintf(stderr, _("Changing password for %s.\n"),
				user = strdup(pwd->pw_name));
		} else {
			fprintf(stderr, _("No user name specified.\n"));
			poptPrintUsage(popt, stderr, 0);
			return 1;
		}
	}

	lu_authenticate_unprivileged(user, "passwd");

	ctx = lu_start(user, groupflag ? lu_group : lu_user, NULL, NULL,
		       interactive ? lu_prompt_console :
		       lu_prompt_console_quiet, NULL, &error);
	if (ctx == NULL) {
		fprintf(stderr, _("Error initializing %s: %s.\n"), PACKAGE,
			lu_strerror(error));
		return 1;
	}

	if ((password == NULL) && (cryptedPassword == NULL) &&
	    (plain_fd == -1) && (crypted_fd == -1)) {
		do {
			struct lu_prompt prompts[2];

			memset(&prompts, 0, sizeof(prompts));
			prompts[0].key = "lpasswd/password1";
			prompts[0].prompt = N_("New password");
			prompts[0].domain = PACKAGE;
			prompts[1].key = "lpasswd/password2";
			prompts[1].prompt = N_("New password (confirm)");
			prompts[1].domain = PACKAGE;
			if (lu_prompt_console(prompts, G_N_ELEMENTS(prompts),
					      NULL, &error)) {
				if (prompts[0].value &&
				    strlen(prompts[0].value) &&
				    prompts[1].value &&
				    strlen(prompts[1].value)) {
					if (strcmp(prompts[0].value,
						   prompts[1].value) == 0) {
						password = g_strdup(prompts[0].value);
						prompts[0].free_value(prompts[0].value);
						prompts[1].free_value(prompts[1].value);
					} else {
						fprintf(stderr, _("Passwords "
							"do not match, try "
							"again.\n"));
					}
				} else {
					fprintf(stderr, _("Password change "
						"canceled.\n"));
					return 1;
				}
			}
			if (error) {
				lu_error_free(&error);
			}
		} while (password == NULL);
	}

	ent = lu_ent_new();

	if (!groupflag) {
		if (lu_user_lookup_name(ctx, user, ent, &error) == FALSE) {
			fprintf(stderr, _("User %s does not exist.\n"), user);
			return 2;
		}
	} else {
		if (lu_group_lookup_name(ctx, user, ent, &error) == FALSE) {
			fprintf(stderr, _("Group %s does not exist.\n"), user);
			return 2;
		}
	}

	if (plain_fd != -1) {
		char buf[LINE_MAX + 1];
		int i;

		i = read(plain_fd, buf, sizeof(buf) - 1);
		while ((i > 0) &&
		       ((buf[i - 1] == '\r') || (buf[i - 1] == '\n')))
			i--;
		buf[i] = '\0';
		password = g_strdup(buf);
		is_crypted = FALSE;
	} else if (crypted_fd != -1) {
		char buf[LINE_MAX + 1];
		int i;
		i = read(crypted_fd, buf, sizeof(buf) - 1);
		while ((i > 0) &&
		       ((buf[i - 1] == '\r') || (buf[i - 1] == '\n')))
			i--;
		buf[i] = '\0';
		password = g_strdup(buf);
		is_crypted = TRUE;
	} else if (cryptedPassword != NULL) {
		password = g_strdup(cryptedPassword);
		is_crypted = TRUE;
	} else {
		is_crypted = FALSE;
		/* the password variable is already set */
	}

	if (!groupflag) {
		if (lu_user_setpass(ctx, ent, password, is_crypted, &error)
		    == FALSE) {
			fprintf(stderr, _("Error setting password for user "
					  "%s: %s.\n"), user,
				lu_strerror(error));
			return 3;
		}
	} else {
		if (lu_group_setpass(ctx, ent, password, is_crypted, &error)
		    == FALSE) {
			fprintf(stderr, _("Error setting password for group "
					  "%s: %s.\n"), user,
				lu_strerror(error));
			return 3;
		}
	}

	lu_ent_free(ent);

	lu_end(ctx);

	fprintf(stderr, _("Password changed.\n"));

	return 0;
}
